dhcp/dhcp_windows.go

```diff
@@ -3,20 +3,181 @@ package dhcp
+	}
+	// Set timeout
+	if err = windows.SetsockoptInt(windows.Handle(fd), windows.SOL_SOCKET, windows.SO_RCVTIMEO, socketTimeoutMillis); err != nil {
+		return ret, errors.Wrap(err, "error setting receive timeout")
+	}
+	return ret, nil
+}
+
+func (s *Socket) Write(packetBytes []byte) (int, error) {
+	err := windows.Sendto(s.fd, packetBytes, 0, &s.destAddr)
+	if err != nil {
+		return 0, errors.Wrap(err, "failed windows send to")
+	}
+	return len(packetBytes), nil
 }
 
-func New(logger *zap.Logger) *DHCP {
-	return &DHCP{
-		logger: logger,
+func (s *Socket) Read(p []byte) (n int, err error) {
+	n, _, innerErr := windows.Recvfrom(s.fd, p, 0)
+	if innerErr != nil {
+		return 0, errors.Wrap(err, "failed windows recv from")
 	}
+	return n, nil
+}
+
+func (s *Socket) Close() error {
+	// do not attempt to close invalid fd (happens on socket creation failure)
+	if s.fd == windows.InvalidHandle {
+		return nil
+	}
+	// Ensure the file descriptor is closed when done
+	if err := windows.Closesocket(s.fd); err != nil {
+		return errors.Wrap(err, "error closing dhcp windows socket")
+	}
+	return nil
 }
 
-func (c *DHCP) DiscoverRequest(_ context.Context, _ net.HardwareAddr, _ string) error {
+// issues a dhcp discover request on an interface by assigning an ip to that interface
+// then, sends a packet with that interface's dummy ip, and then unassigns the dummy ip
+func (c *DHCP) DiscoverRequest(ctx context.Context, macAddress net.HardwareAddr, ifName string) error {
+	// validate interface name
+	if !allowedInput.MatchString(ifName) {
+		return errors.New("invalid dhcp discover request interface name")
+	}
+	// delete dummy ip off the interface if it already exists
+	ret, err := c.execClient.ExecuteCommand(ctx, "netsh", "interface", "ipv4", "delete", "address", ifName, dummyIPAddressStr)
+	if err != nil {
+		c.logger.Info("Could not remove dummy ip, likely because it doesn't exist", zap.String("output", ret), zap.Error(log.NewErrorWithoutStackTrace(err)))
+	}
+	time.Sleep(deleteIPAddressDelay)
+
+	// create dummy ip so we can direct the packet to the correct interface
+	ret, err = c.execClient.ExecuteCommand(ctx, "netsh", "interface", "ipv4", "add", "address", ifName, dummyIPAddressStr, dummySubnetMask)
+	if err != nil {
+		return errors.Wrap(err, "failed to add dummy ip to interface: "+ret)
+	}
+	// ensure we always remove the dummy ip we added from the interface
+	defer func() {
+		// we always want to try to remove the dummy ip, even if the deadline was reached
+		// so we have context.Background()
+		ret, cleanupErr := c.execClient.ExecuteCommand(context.Background(), "netsh", "interface", "ipv4", "delete", "address", ifName, dummyIPAddressStr)
+		if cleanupErr != nil {
+			c.logger.Info("Failed to remove dummy ip on leaving function", zap.String("output", ret), zap.Error(err))
+		}
+		// wait for nic to retrieve autoconfiguration ip
+		time.Sleep(returnDelay)
+	}()
+	// it takes time for the address to be assigned
+	time.Sleep(addIPAddressDelay)
```

Comment:
Don't we have a way to test for the new ip to be assigned instead of guessing a delay?

Reply:
Yeah, +1 to @estebancams . I'd rather have a fast check loop here (500ms sleep) then time out.